Add Kept on Death plugin

TheStonedTurtle opened the pull request:
Modifies the Items Kept on Death widget to interactive and slightly more accurate. If the player is inside a safe zone the normal interface will be shown.

### Although this is an improvement on the original interface there may be situations where the new interface is wrong. Please report these cases so we can improve the interface further

![2018-11-25_20-35-31](https://user-images.githubusercontent.com/29030969/48993323-e4ba4f80-f0f1-11e8-9da0-ae28c152af49.gif)

It will automatically detect if the player is Skulled, Protecting Item, and the Wilderness safety (Low Lvl/High). It also support PvP world Safe/Non Safe areas.

## Death Mechanics

<details>
<p>

### Known Death Mechanics:
These mechanics should never change.

1) Every account protects 3 items.
2) Skulled account loses all Items
3) Protect item keeps an additional item based on value
4) UIM accounts lose all items, like they were Skulled.

### Assumed Death Mechanics:
These mechanics were implemented due to server instability and are still not fully understood.

1) **Outside wildy (Non-PvP death)**: Most non-tradeable items are kept (ie, Looting bag is lost)
2) **Lvl 1-20 wildy**: Non-tradeable items are lost on death unless they are one that turns into a broken variant.
3) **Above 20 wildy**: Non-tradeable items are lost on death and turned into coins.

The assumed death mechanics apply after the Known death mechanics which means You can keep untradeable items above 20 wildy if they are kept by the Known death mechanics.

</p>
</details>

raiyni commented on runelite-client/src/main/java/net/runelite/client/plugins/keptondeath/ActuallyTradeableItem.java:
```diff
@@ -0,0 +1,83 @@
+
+	public static Boolean check(int id)
+	{
+		return ID_MAP.get(id) != null;
```
use a hash set

raiyni commented on runelite-client/src/main/java/net/runelite/client/plugins/keptondeath/BrokenOnDeathItem.java:
```diff
@@ -0,0 +1,109 @@
+
+	public static boolean check(int itemID)
+	{
+		return ID_MAP.get(itemID) != null;
```
use a hash set

deathbeam commented:
Imo this could be placed inside the death indicator plugin, not sure

daviscook477 commented:
The Updated Skulled reminder text doesn't seem too helpful. Why remove the bit about "the normal amount of items you keep is 3. You're marked with a PK skull ....". And the bit that says "Protecting 0 items by default PK skull -3" isn't really all that clear.

daviscook477 commented:
Also I'm not too familiar with the wilderness mechanics but why would the bag looking thing with the white outline be lost in <20 wilderness but kept when skulled. Is that right?

TheStonedTurtle commented:
@daviscook477 The Rune Pouch has a white outline by default but is protected on deaths in non-pvp scenarios. The skulled pictures were in a safe zone so it was added to the top part without the white outline. I guess I could remove the white outline in both cases.

As for why I updated the text, its extremely cluttered with the default text and I added two new sentences which would not have fit otherwise. This photo doesn't even have the bond text.

![https://i.imgur.com/0T7UM1z.png](https://i.imgur.com/0T7UM1z.png)

daviscook477 commented:
Maybe removing the white outline would be helpful. It's a bit misleading for the game to say "items with a white outline will always be lost" and then it not be true. I still think the "Protecting 0 items by default PK skull -3" is a bit confusing. What is that text trying to communicate?

TheStonedTurtle commented:
The interface has been reworked to be interactive with toggles for the different effects. See gif:
![2018-11-25_20-35-31](https://user-images.githubusercontent.com/29030969/48993323-e4ba4f80-f0f1-11e8-9da0-ae28c152af49.gif)

A few use cases were also addressed:
* Added PvP world support
* Show default interface inside safe zones (PoH/Clan Wars/etc)
* Fixed incorrect interface being shown when widget is rebuilt

TheStonedTurtle commented:
Value of items doesn't work correctly for noted items
Max items kept on death doesn't account for stacked items

![https://i.imgur.com/GFRNa9R.png](https://i.imgur.com/GFRNa9R.png)

Looks like rune pouch doesn't count towards total but does currently? Not sure about the other AlwaysLost items

![https://i.imgur.com/oPsJQna.png](https://i.imgur.com/oPsJQna.png)

Extra blank row for kept items? (Barrows is degraded btw)

![https://i.imgur.com/BwboVik.png](https://i.imgur.com/BwboVik.png)

TheStonedTurtle commented:
I fixed a few minor things, but I'm unsure of how the game handles ornate items in the wilderness. I.E. occult ornate is 4.8m (3.9m kit, 900k occult). What value does the game use?

![https://i.imgur.com/XhAPACP.png](https://i.imgur.com/XhAPACP.png)

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/itemskeptondeath/AlwaysLostItem.java:
```diff
@@ -0,0 +1,65 @@
+	CLUE_BOX(ItemID.CLUE_BOX, false);
+
+	private final int itemID;
+	private final boolean kept;
```
This should be named keptOutsideOfWilderness, or at least documented like that, otherwise it is very confusing to have enum named "AlwaysLostItem" and then field inside named "kept". I had to actually go through the plugin implementation to understand what this field means.

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/itemskeptondeath/AlwaysLostItem.java:
```diff
@@ -0,0 +1,65 @@
+		ID_MAP = map.build();
+	}
+
+	static AlwaysLostItem getByItemID(final int itemID)
```
We should keep some consistent naming for these methods, e.g clue scroll plugin is using forItemId

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/itemskeptondeath/BrokenOnDeathItem.java:
```diff
@@ -0,0 +1,111 @@
+		ID_SET = set.build();
+	}
+
+	static boolean check(final int itemID)
```
check?

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/itemskeptondeath/ItemsKeptOnDeathPlugin.java:
```diff
@@ -0,0 +1,582 @@
+	 * @param c   Items Composition
+	 * @return
+	 */
+	private Widget createItemWidget(final Widget parent, final int qty, final ItemComposition c)
```
static

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/itemskeptondeath/ItemsKeptOnDeathPlugin.java:
```diff
@@ -0,0 +1,582 @@
+	 * @param c The item
+	 * @return
+	 */
+	private boolean isTradeable(final ItemComposition c)
```
static

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/itemskeptondeath/ItemsKeptOnDeathPlugin.java:
```diff
@@ -0,0 +1,582 @@
+	/**
+	 * Position a list of widget items in the parent container
+	 */
+	private void positionWidgetItems(final Widget parent, final List<Widget> widgets)
```
static

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/itemskeptondeath/ItemsKeptOnDeathPlugin.java:
```diff
@@ -0,0 +1,582 @@
+	private int wildyLevel;
+
+	@Subscribe
+	protected void onScriptCallbackEvent(ScriptCallbackEvent event)
```
We use private everywhere else atm

Adam- replied:
I think you mean public?

deathbeam replied:
right, yes :d
